cmake/llvm_setup.cmake

```
@@ -0,0 +1,116 @@
+include_guard()
+
+function(fetch_private_clang_files)
+    # files to fetch:
+    # clang/lib/Sema/CoroutineStmtBuilder.h
+    # clang/lib/Sema/TypeLocBuilder.h
+    # clang/lib/Sema/TreeTransform.h
+    # commit hash 7b09d7b446383b71b63d429b21ee45ba389c5134
+
+    # check private clang files' presence
+    file(GLOB_RECURSE CLANG_PRIVATE_FILES "${LLVM_INSTALL_PATH}/include/clang/Sema/TreeTransform.h")
+    if(CLANG_PRIVATE_FILES)
+        message(STATUS "Found private clang files presents, skip downloading.")
+        return()
+    endif()
+
+    message(WARNING "Private clang files not found, try fetch from llvm-project source...")
+
+    file(MAKE_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/include/clang/lib/Sema")
+    file(
+        DOWNLOAD "https://raw.githubusercontent.com/llvm/llvm-project/7b09d7b446383b71b63d429b21ee45ba389c5134/clang/lib/Sema/CoroutineStmtBuilder.h" "${CMAKE_CURRENT_BINARY_DIR}/include/clang/Sema/CoroutineStmtBuilder.h"
+        SHOW_PROGRESS
+    )
+    file(
+        DOWNLOAD "https://raw.githubusercontent.com/llvm/llvm-project/7b09d7b446383b71b63d429b21ee45ba389c5134/clang/lib/Sema/TypeLocBuilder.h" "${CMAKE_CURRENT_BINARY_DIR}/include/clang/Sema/TypeLocBuilder.h"
+        SHOW_PROGRESS
+    )
+    file(
+        DOWNLOAD "https://raw.githubusercontent.com/llvm/llvm-project/7b09d7b446383b71b63d429b21ee45ba389c5134/clang/lib/Sema/TreeTransform.h" "${CMAKE_CURRENT_BINARY_DIR}/include/clang/Sema/TreeTransform.h"
+        SHOW_PROGRESS
+    )
+
+    # add to include directories
+    target_include_directories(llvm-libs INTERFACE "${CMAKE_CURRENT_BINARY_DIR}/include")
+endfunction()
+
+function(detect_llvm result)
+    find_program(LLVM_CONFIG_EXEC llvm-config)
+    if(NOT LLVM_CONFIG_EXEC)
+        set(${result} FALSE PARENT_SCOPE)
+        return()
+    endif()
+    execute_process(
+        COMMAND "${LLVM_CONFIG_EXEC}" --version
+        OUTPUT_VARIABLE LLVM_VERSION
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+    )
+    if(NOT LLVM_VERSION VERSION_EQUAL "20.1.5")
```

```suggestion
    if(NOT (LLVM_VERSION VERSION_GREATER_EQUAL "20.1" AND LLVM_VERSION VERSION_LESS_EQUAL "20.2"))
```
Major version part is enough although.